File: python/src/frtb/sbm.py
"""Sensitivities-Based Method (SBM) aggregation.

Formulas (FRTB structure, educational parameter set):

Weighted sensitivity            WS_k = RW_k * s_k
Within-bucket                   K_b  = sqrt(max(0, sum_k WS_k^2
                                        + sum_{k != l} rho_kl WS_k WS_l))
Across buckets                  Charge = sqrt(max(0, sum_b K_b^2
                                        + sum_{b != c} gamma_bc S_b S_c))
  with S_b = sum_k WS_k; if the argument of the outer sqrt is negative the
  S_b FALLBACK applies:  S_b = max(min(sum_k WS_k, K_b), -K_b)  and the
  aggregate is recomputed (the max(0, .) guard is kept as a belt-and-braces
  guard against negative rounding).

Correlation scenarios: for every rho and gamma,
  high   rho -> min(1.25 * rho, 1.0)
  medium rho -> rho
  low    rho -> 0.75 * rho          (pinned simplification of Basel's
                                     max(2*rho - 1, 0.75*rho))
The risk-class charges are computed under each scenario, summed across risk
classes and the SBM capital is the MAX of the three scenario totals.

Curvature: per curvature risk factor k with delta sensitivity s_k,
  CVR_k^+ = -( V(x_k up)   - V - RW_k^curv * s_k )
  CVR_k^- = -( V(x_k down) - V + RW_k^curv * s_k )
  K_b^+/- = sqrt(max(0, max(sum-of-squares form with psi) ))  where
            psi(a, b) = 0 if a < 0 and b < 0 else 1
  K_b     = max(K_b^+, K_b^-), select S_b as the CVR sum of the winning side
            (ties -> up side; pinned simplification, documented in API_SPEC).
Across buckets curvature uses gamma^2, psi(S_b, S_c) and max(0, .) inside the
sqrt; there is no S_b fallback for curvature. Curvature correlations are the
squares of the (scenario-scaled) delta correlations (pinned simplification).
"""
from __future__ import annotations

import math
from dataclasses import dataclass, field
from typing import Callable, Dict, List, Mapping, Sequence, Tuple

SCENARIOS = ("high", "medium", "low")


def scale_rho(rho: float, scenario: str, high: float = 1.25, low: float = 0.75) -> float:
    """Apply the correlation scenario scaler; 'high' is capped at 1.0."""
    if scenario == "medium":
        return rho
    if scenario == "high":
        return min(high * rho, 1.0)
    if scenario == "low":
        return low * rho
    raise ValueError(f"scale_rho: unknown scenario '{scenario}'")


def bucket_kb(ws: Sequence[float], rho: Callable[[int, int], float]) -> float:
    """Within-bucket charge K_b = sqrt(max(0, sum WS^2 + sum_{k!=l} rho WS_k WS_l)).

    ``rho(k, l)`` supplies the pairwise correlation for k != l.
    The max(0, .) guard protects against negative rounding of the quadratic form.
    """
    for w in ws:
        if not math.isfinite(w):
            raise ValueError("bucket_kb: weighted sensitivities must be finite")
    total = sum(w * w for w in ws)
    n = len(ws)
    for k in range(n):
        for l in range(n):
            if k != l:
                total += rho(k, l) * ws[k] * ws[l]
    return math.sqrt(max(0.0, total))


@dataclass(frozen=True)
class AggregateResult:
    """Across-bucket aggregation output (with the S_b fallback bookkeeping)."""

    charge: float
    used_fallback: bool
    sb: Dict[str, float]  # the S_b actually used (post-fallback if triggered)


def aggregate_buckets(
    kb: Mapping[str, float],
    ws_sum: Mapping[str, float],
    gamma: Callable[[str, str], float],
) -> AggregateResult:
    """Across-bucket aggregation with the FRTB S_b fallback rule.

    First tries S_b = sum_k WS_k.  If sum_b K_b^2 + sum_{b!=c} gamma S_b S_c
    < 0, recomputes with S_b = max(min(sum_k WS_k, K_b), -K_b).
    """
    if set(kb) != set(ws_sum):
        raise ValueError("aggregate_buckets: kb and ws_sum must cover the same buckets")
    buckets = sorted(kb)

    def _inner(sb: Mapping[str, float]) -> float:
        total = sum(kb[b] ** 2 for b in buckets)
        for b in buckets:
            for c in buckets:
                if b != c:
                    total += gamma(b, c) * sb[b] * sb[c]
        return total

    sb0 = {b: ws_sum[b] for b in buckets}
    inner = _inner(sb0)
    if inner >= 0.0:
        return AggregateResult(math.sqrt(inner), False, sb0)
    sb1 = {b: max(min(ws_sum[b], kb[b]), -kb[b]) for b in buckets}
    inner = _inner(sb1)
    return AggregateResult(math.sqrt(max(0.0, inner)), True, sb1)


# --------------------------------------------------------------------------
# Delta / vega charges
# --------------------------------------------------------------------------

@dataclass(frozen=True)
class RiskClassCharge:
    """Per-scenario charge with per-bucket K_b detail (for reporting/golden)."""

    charge: float
    kb: Dict[str, float]
    used_fallback: bool


def delta_vega_charge(
    bucket_ws: Mapping[str, Mapping[str, float]],
    intra_rho: Callable[[str, str, str], float],
    gamma: Callable[[str, str], float],
    scenario: str,
    scenario_high: float = 1.25,
    scenario_low: float = 0.75,
) -> RiskClassCharge:
    """Generic delta or vega charge for one risk class under one scenario.

    bucket_ws:  {bucket: {factor_key: WS}}
    intra_rho:  (bucket, factor_k, factor_l) -> medium-scenario correlation
    gamma:      (bucket_b, bucket_c) -> medium-scenario cross-bucket gamma
    """
    kb: Dict[str, float] = {}
    ws_sum: Dict[str, float] = {}
    for b, factors in bucket_ws.items():
        keys = sorted(factors)
        ws = [factors[k] for k in keys]

        def rho_fn(i: int, j: int, _b: str = b, _keys: List[str] = keys) -> float:
            return scale_rho(intra_rho(_b, _keys[i], _keys[j]), scenario,
                             scenario_high, scenario_low)

        kb[b] = bucket_kb(ws, rho_fn)
        ws_sum[b] = sum(ws)

    if not kb:
        return RiskClassCharge(0.0, {}, False)

    def gamma_fn(b: str, c: str) -> float:
        return scale_rho(gamma(b, c), scenario, scenario_high, scenario_low)

    agg = aggregate_buckets(kb, ws_sum, gamma_fn)
    return RiskClassCharge(agg.charge, kb, agg.used_fallback)


# --------------------------------------------------------------------------
# Curvature
# --------------------------------------------------------------------------

def psi(a: float, b: float) -> float:
    """FRTB psi: 0 when both CVR terms are negative, else 1."""
    return 0.0 if (a < 0.0 and b < 0.0) else 1.0


def curvature_bucket_kb(
    cvr_up: Sequence[float],
    cvr_dn: Sequence[float],
    rho: Callable[[int, int], float],
) -> Tuple[float, float]:
    """Within-bucket curvature charge.

    Returns (K_b, S_b) where K_b = max(K_b+, K_b-) with
      K_b+/- = sqrt(max(0, sum_k max(CVR_k,0)^2
                         + sum_{k!=l} rho_kl CVR_k CVR_l psi(CVR_k, CVR_l)))
    and S_b is the sum of CVRs on the winning side (up on ties).
    ``rho`` must already be the CURVATURE correlation (delta rho squared).
    """
    if len(cvr_up) != len(cvr_dn):
        raise ValueError("curvature_bucket_kb: up/down CVR lists must match")

    def side(cvr: Sequence[float]) -> float:
        total = sum(max(c, 0.0) ** 2 for c in cvr)
        n = len(cvr)
        for k in range(n):
            for l in range(n):
                if k != l:
                    total += rho(k, l) * cvr[k] * cvr[l] * psi(cvr[k], cvr[l])
        return math.sqrt(max(0.0, total))

    k_up, k_dn = side(cvr_up), side(cvr_dn)
    if k_up >= k_dn:
        return k_up, sum(cvr_up)
    return k_dn, sum(cvr_dn)


def curvature_charge(
    bucket_cvr: Mapping[str, Tuple[Sequence[float], Sequence[float]]],
    intra_rho: Callable[[str, str, str], float],
    gamma: Callable[[str, str], float],
    scenario: str,
    factor_keys: Mapping[str, Sequence[str]],
    scenario_high: float = 1.25,
    scenario_low: float = 0.75,
) -> RiskClassCharge:
    """Curvature charge for one risk class under one scenario.

    bucket_cvr: {bucket: (list CVR+, list CVR-)} aligned with factor_keys[bucket].
    intra_rho / gamma supply medium DELTA correlations; they are scenario-scaled
    then SQUARED for curvature (pinned simplification, documented).
    """
    kb: Dict[str, float] = {}
    sb: Dict[str, float] = {}
    for b, (up, dn) in bucket_cvr.items():
        keys = list(factor_keys[b])
        if len(keys) != len(up):
            raise ValueError(f"curvature_charge: factor keys mismatch in bucket '{b}'")

        def rho_fn(i: int, j: int, _b: str = b, _keys: List[str] = keys) -> float:
            r = scale_rho(intra_rho(_b, _keys[i], _keys[j]), scenario,
                          scenario_high, scenario_low)
            return r * r

        kb[b], sb[b] = curvature_bucket_kb(up, dn, rho_fn)

    if not kb:
        return RiskClassCharge(0.0, {}, False)

    buckets = sorted(kb)
    total = sum(kb[b] ** 2 for b in buckets)
    for b in buckets:
        for c in buckets:
            if b != c:
                g = scale_rho(gamma(b, c), scenario, scenario_high, scenario_low)
                total += (g * g) * sb[b] * sb[c] * psi(sb[b], sb[c])
    return RiskClassCharge(math.sqrt(max(0.0, total)), kb, False)

File: java/src/main/java/com/quant/frtb/Sbm.java
package com.quant.frtb;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * Sensitivities-Based Method (SBM) aggregation.
 *
 * <p>Formulas (FRTB structure, educational parameter set):
 * <pre>
 * Weighted sensitivity   WS_k = RW_k * s_k
 * Within-bucket          K_b  = sqrt(max(0, sum WS^2 + sum_{k!=l} rho_kl WS_k WS_l))
 * Across buckets         Charge = sqrt(max(0, sum K_b^2 + sum_{b!=c} gamma_bc S_b S_c))
 * </pre>
 * with {@code S_b = sum_k WS_k}; if the argument of the outer sqrt is
 * negative the S_b FALLBACK applies: {@code S_b = max(min(sum WS_k, K_b), -K_b)}
 * and the aggregate is recomputed once (the max(0,.) guard is kept).
 *
 * <p>Correlation scenarios (applied to every rho and gamma): high
 * {@code min(1.25*rho, 1)}, medium {@code rho}, low {@code 0.75*rho} (pinned
 * simplification of Basel's {@code max(2*rho-1, 0.75*rho)}). The SBM capital
 * is the max over the three scenario totals.
 *
 * <p>Curvature uses CVR+/CVR- per factor, the psi both-negative rule at both
 * aggregation levels, squared (scenario-scaled) delta correlations, no S_b
 * fallback, and up-side selection on K_b+ = K_b- ties (pinned
 * simplifications, documented in API_SPEC.md).
 */
public final class Sbm {

    private Sbm() {
    }

    /** The three correlation scenarios, in reporting order. */
    public static final List<String> SCENARIOS = List.of("high", "medium", "low");

    /** Pairwise correlation lookup by factor index (k != l). */
    @FunctionalInterface
    public interface PairCorr {
        /** Correlation between factors k and l of a bucket. */
        double get(int k, int l);
    }

    /** Medium-scenario intra-bucket correlation lookup by bucket and factor keys. */
    @FunctionalInterface
    public interface IntraRho {
        /** Correlation between factor keys k and l inside a bucket. */
        double get(String bucket, String k, String l);
    }

    /** Medium-scenario cross-bucket gamma lookup. */
    @FunctionalInterface
    public interface CrossGamma {
        /** Gamma between buckets b and c. */
        double get(String b, String c);
    }

    /** Apply the correlation scenario scaler; "high" is capped at 1.0. */
    public static double scaleRho(double rho, String scenario, double high, double low) {
        switch (scenario) {
            case "medium":
                return rho;
            case "high":
                return Math.min(high * rho, 1.0);
            case "low":
                return low * rho;
            default:
                throw new IllegalArgumentException("scaleRho: unknown scenario '" + scenario + "'");
        }
    }

    /** {@code scaleRho} with the pinned default scalers 1.25 / 0.75. */
    public static double scaleRho(double rho, String scenario) {
        return scaleRho(rho, scenario, 1.25, 0.75);
    }

    /**
     * Within-bucket charge
     * {@code K_b = sqrt(max(0, sum WS^2 + sum_{k!=l} rho WS_k WS_l))}.
     *
     * <p>The max(0,.) guard protects against negative rounding of the
     * quadratic form.
     */
    public static double bucketKb(List<Double> ws, PairCorr rho) {
        for (double w : ws) {
            if (!Double.isFinite(w)) {
                throw new IllegalArgumentException(
                        "bucketKb: weighted sensitivities must be finite");
            }
        }
        double total = 0.0;
        for (double w : ws) {
            total += w * w;
        }
        int n = ws.size();
        for (int k = 0; k < n; k++) {
            for (int l = 0; l < n; l++) {
                if (k != l) {
                    total += rho.get(k, l) * ws.get(k) * ws.get(l);
                }
            }
        }
        return Math.sqrt(Math.max(0.0, total));
    }

    /**
     * Across-bucket aggregation output (with the S_b fallback bookkeeping).
     *
     * @param charge       the aggregated charge
     * @param usedFallback true when the S_b fallback branch was taken
     * @param sb           the S_b actually used (post-fallback if triggered)
     */
    public record AggregateResult(double charge, boolean usedFallback, Map<String, Double> sb) {
    }

    /**
     * Across-bucket aggregation with the FRTB S_b fallback rule.
     *
     * <p>First tries {@code S_b = sum_k WS_k}. If
     * {@code sum K_b^2 + sum_{b!=c} gamma S_b S_c < 0}, recomputes once with
     * {@code S_b = max(min(sum_k WS_k, K_b), -K_b)}.
     */
    public static AggregateResult aggregateBuckets(Map<String, Double> kb,
                                                   Map<String, Double> wsSum,
                                                   CrossGamma gamma) {
        if (!kb.keySet().equals(wsSum.keySet())) {
            throw new IllegalArgumentException(
                    "aggregateBuckets: kb and wsSum must cover the same buckets");
        }
        List<String> buckets = new ArrayList<>(kb.keySet());
        Collections.sort(buckets);

        Map<String, Double> sb0 = new LinkedHashMap<>();
        for (String b : buckets) {
            sb0.put(b, wsSum.get(b));
        }
        double inner = aggregateInner(buckets, kb, sb0, gamma);
        if (inner >= 0.0) {
            return new AggregateResult(Math.sqrt(inner), false, sb0);
        }
        Map<String, Double> sb1 = new LinkedHashMap<>();
        for (String b : buckets) {
            sb1.put(b, Math.max(Math.min(wsSum.get(b), kb.get(b)), -kb.get(b)));
        }
        inner = aggregateInner(buckets, kb, sb1, gamma);
        return new AggregateResult(Math.sqrt(Math.max(0.0, inner)), true, sb1);
    }

    private static double aggregateInner(List<String> buckets, Map<String, Double> kb,
                                         Map<String, Double> sb, CrossGamma gamma) {
        double total = 0.0;
        for (String b : buckets) {
            double k = kb.get(b);
            total += k * k;
        }
        for (String b : buckets) {
            for (String c : buckets) {
                if (!b.equals(c)) {
                    total += gamma.get(b, c) * sb.get(b) * sb.get(c);
                }
            }
        }
        return total;
    }

    // ---------------------------------------------------------------------
    // Delta / vega charges
    // ---------------------------------------------------------------------

    /**
     * Per-scenario charge with per-bucket K_b detail (for reporting/golden).
     *
     * @param charge       the risk-class charge under one scenario
     * @param kb           per-bucket within-bucket charges
     * @param usedFallback whether the across-bucket S_b fallback fired
     */
    public record RiskClassCharge(double charge, Map<String, Double> kb, boolean usedFallback) {
    }

    /**
     * Generic delta or vega charge for one risk class under one scenario.
     *
     * @param bucketWs  {@code bucket -> factorKey -> WS}
     * @param intraRho  (bucket, factorK, factorL) -&gt; medium-scenario correlation
     * @param gamma     (bucketB, bucketC) -&gt; medium-scenario cross-bucket gamma
     * @param scenario  "high" | "medium" | "low"
     * @param scenarioHigh high scaler (1.25 pinned)
     * @param scenarioLow  low scaler (0.75 pinned)
     * @return the charge with per-bucket K_b detail
     */
    public static RiskClassCharge deltaVegaCharge(Map<String, Map<String, Double>> bucketWs,
                                                  IntraRho intraRho, CrossGamma gamma,
                                                  String scenario,
                                                  double scenarioHigh, double scenarioLow) {
        Map<String, Double> kb = new LinkedHashMap<>();
        Map<String, Double> wsSum = new LinkedHashMap<>();
        for (Map.Entry<String, Map<String, Double>> e : bucketWs.entrySet()) {
            String b = e.getKey();
            List<String> keys = new ArrayList<>(e.getValue().keySet());
            Collections.sort(keys);
            List<Double> ws = new ArrayList<>();
            for (String key : keys) {
                ws.add(e.getValue().get(key));
            }
            PairCorr rhoFn = (i, j) -> scaleRho(intraRho.get(b, keys.get(i), keys.get(j)),
                    scenario, scenarioHigh, scenarioLow);
            kb.put(b, bucketKb(ws, rhoFn));
            double sum = 0.0;
            for (double w : ws) {
                sum += w;
            }
            wsSum.put(b, sum);
        }

        if (kb.isEmpty()) {
            return new RiskClassCharge(0.0, Map.of(), false);
        }
        CrossGamma gammaFn = (b, c) -> scaleRho(gamma.get(b, c), scenario,
                scenarioHigh, scenarioLow);
        AggregateResult agg = aggregateBuckets(kb, wsSum, gammaFn);
        return new RiskClassCharge(agg.charge(), kb, agg.usedFallback());
    }

    // ---------------------------------------------------------------------
    // Curvature
    // ---------------------------------------------------------------------

    /** FRTB psi: 0 when both CVR terms are negative, else 1. */
    public static double psi(double a, double b) {
        return (a < 0.0 && b < 0.0) ? 0.0 : 1.0;
    }

    /**
     * Within-bucket curvature charge and side selection.
     *
     * @param kb the winning K_b = max(K_b+, K_b-)
     * @param sb the CVR sum on the winning side (up side on ties)
     */
    public record KbSb(double kb, double sb) {
    }

    /**
     * Within-bucket curvature charge.
     *
     * <p>{@code K_b+/- = sqrt(max(0, sum max(CVR,0)^2 + sum_{k!=l} rho CVR_k CVR_l psi))},
     * K_b = max of the sides, S_b = sum of CVRs on the winning side (up on
     * ties). {@code rho} must already be the CURVATURE correlation (the
     * scenario-scaled delta rho squared).
     */
    public static KbSb curvatureBucketKb(List<Double> cvrUp, List<Double> cvrDn, PairCorr rho) {
        if (cvrUp.size() != cvrDn.size()) {
            throw new IllegalArgumentException("curvatureBucketKb: up/down CVR lists must match");
        }
        double kUp = curvatureSide(cvrUp, rho);
        double kDn = curvatureSide(cvrDn, rho);
        if (kUp >= kDn) {
            return new KbSb(kUp, sum(cvrUp));
        }
        return new KbSb(kDn, sum(cvrDn));
    }

    private static double curvatureSide(List<Double> cvr, PairCorr rho) {
        double total = 0.0;
        for (double c : cvr) {
            double m = Math.max(c, 0.0);
            total += m * m;
        }
        int n = cvr.size();
        for (int k = 0; k < n; k++) {
            for (int l = 0; l < n; l++) {
                if (k != l) {
                    total += rho.get(k, l) * cvr.get(k) * cvr.get(l)
                            * psi(cvr.get(k), cvr.get(l));
                }
            }
        }
        return Math.sqrt(Math.max(0.0, total));
    }

    private static double sum(List<Double> xs) {
        double s = 0.0;
        for (double x : xs) {
            s += x;
        }
        return s;
    }

    /**
     * Curvature charge for one risk class under one scenario.
     *
     * @param bucketCvr  {@code bucket -> [CVR+ list, CVR- list]} aligned with
     *                   {@code factorKeys.get(bucket)}
     * @param intraRho   medium DELTA correlations; scenario-scaled then SQUARED
     *                   for curvature (pinned simplification)
     * @param gamma      medium DELTA cross-bucket gamma, same treatment
     * @param scenario   "high" | "medium" | "low"
     * @param factorKeys per-bucket factor key lists for the rho lookups
     * @param scenarioHigh high scaler
     * @param scenarioLow  low scaler
     * @return the curvature charge with per-bucket K_b detail
     */
    public static RiskClassCharge curvatureCharge(Map<String, List<List<Double>>> bucketCvr,
                                                  IntraRho intraRho, CrossGamma gamma,
                                                  String scenario,
                                                  Map<String, List<String>> factorKeys,
                                                  double scenarioHigh, double scenarioLow) {
        Map<String, Double> kb = new LinkedHashMap<>();
        Map<String, Double> sb = new LinkedHashMap<>();
        for (Map.Entry<String, List<List<Double>>> e : bucketCvr.entrySet()) {
            String b = e.getKey();
            List<Double> up = e.getValue().get(0);
            List<Double> dn = e.getValue().get(1);
            List<String> keys = factorKeys.get(b);
            if (keys == null || keys.size() != up.size()) {
                throw new IllegalArgumentException(
                        "curvatureCharge: factor keys mismatch in bucket '" + b + "'");
            }
            PairCorr rhoFn = (i, j) -> {
                double r = scaleRho(intraRho.get(b, keys.get(i), keys.get(j)), scenario,
                        scenarioHigh, scenarioLow);
                return r * r;
            };
            KbSb res = curvatureBucketKb(up, dn, rhoFn);
            kb.put(b, res.kb());
            sb.put(b, res.sb());
        }

        if (kb.isEmpty()) {
            return new RiskClassCharge(0.0, Map.of(), false);
        }
        List<String> buckets = new ArrayList<>(kb.keySet());
        Collections.sort(buckets);
        double total = 0.0;
        for (String b : buckets) {
            double k = kb.get(b);
            total += k * k;
        }
        for (String b : buckets) {
            for (String c : buckets) {
                if (!b.equals(c)) {
                    double g = scaleRho(gamma.get(b, c), scenario, scenarioHigh, scenarioLow);
                    total += (g * g) * sb.get(b) * sb.get(c) * psi(sb.get(b), sb.get(c));
                }
            }
        }
        return new RiskClassCharge(Math.sqrt(Math.max(0.0, total)), kb, false);
    }
}
